File: backend-spring/application/src/main/java/apptemplate/application/mappers/UserMapper.java
package apptemplate.application.mappers;

import apptemplate.application.dto.auth.UserInfoResponse;
import apptemplate.application.dto.user.UserDto;
import apptemplate.domain.entities.User;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.factory.Mappers;

@Mapper(componentModel = "spring")
public interface UserMapper {

    UserMapper INSTANCE = Mappers.getMapper(UserMapper.class);

    @Mapping(target = "role", expression = "java(user.getRole().name())")
    @Mapping(target = "departmentName", ignore = true)
    @Mapping(target = "fullName", source = "name")
    @Mapping(target = "isActive", source = "active")
    @Mapping(target = "firstName", expression = "java(user.getName() != null ? user.getName().split(\" \")[0] : \"\")")
    @Mapping(target = "lastName", expression = "java(user.getName() != null && user.getName().contains(\" \") ? user.getName().substring(user.getName().indexOf(\" \") + 1) : \"\")")
    UserDto toDto(User user);

    @Mapping(target = "role", expression = "java(user.getRole().name())")
    @Mapping(target = "departmentName", ignore = true)
    @Mapping(target = "fullName", source = "name")
    @Mapping(target = "isActive", source = "active")
    @Mapping(target = "firstName", expression = "java(user.getName() != null ? user.getName().split(\" \")[0] : \"\")")
    @Mapping(target = "lastName", expression = "java(user.getName() != null && user.getName().contains(\" \") ? user.getName().substring(user.getName().indexOf(\" \") + 1) : \"\")")
    UserInfoResponse toUserInfoResponse(User user);
}

File: backend/nestjs/clean-architecture/full/src/modules/user-management/application/mappers/user.mapper.ts
import { User } from '../../domain/entities/user.entity';
import { UserDto } from '../dto/user.dto';

export class UserMapper {
  static toDto(user: User, departmentName?: string | null): UserDto {
    return new UserDto({
      id: user.id,
      email: user.email,
      username: user.username,
      firstName: user.firstName,
      lastName: user.lastName,
      fullName: user.fullName,
      role: user.role,
      departmentId: user.departmentId,
      departmentName: departmentName ?? null,
      isActive: user.isActive,
      lastLoginAt: user.lastLoginAt,
      createdAt: user.createdAt,
      updatedAt: user.updatedAt,
    });
  }

  static toDtoList(users: User[]): UserDto[] {
    return users.map((user) => UserMapper.toDto(user));
  }
}
